### ISSUE TITLE
AttributeError: Can't pickle local object 'Optimizer.<locals>._GradientAccumulationOptimizer.State'

### ISSUE THREAD

**Issue by sandeepnmenon** [run-ai/runai#28]
I am using the runai.ga.torch module to define my optimizer as follows
```
optimizer = torch.optim.SGD(
            model.parameters(),
            lr=configs.optimizer.lr,
            momentum=configs.optimizer.momentum,
            weight_decay=configs.optimizer.weight_decay,
            nesterov=configs.optimizer.nesterov)
import runai.ga.torch
optimizer = runai.ga.torch.optim.Optimizer(optimizer, 2)
```
While saving the model after the epoch I am getting the following error
```
  File "/home/deepenai/SandeepMenon/venv-e3d/lib/python3.6/site-packages/torchpack/utils/io.py", line 99, in save_pt
    torch.save(obj, f, **kwargs)
  File "/home/deepenai/SandeepMenon/venv-e3d/lib/python3.6/site-packages/torch/serialization.py", line 372, in save
    _save(obj, opened_zipfile, pickle_module, pickle_protocol)
  File "/home/deepenai/SandeepMenon/venv-e3d/lib/python3.6/site-packages/torch/serialization.py", line 476, in _save
    pickler.dump(obj)
AttributeError: Can't pickle local object 'Optimizer.<locals>._GradientAccumulationOptimizer.State'
```

These are my state_dict values
```
state_dict = dict()
state_dict['model'] = self.model.state_dict()
state_dict['optimizer'] = self.optimizer.state_dict()
state_dict['scheduler'] = self.scheduler.state_dict()
state_dict['callbacks'] = self.callbacks.state_dict()
state_dict['epoch_num'] = self.epoch_num
state_dict['local_step'] = self.local_step
state_dict['global_step'] = self.global_step
```

My pipeline worked fine before changing my optimizer to runai.ga.torch.optim.Optimizer, so the optimizer.state_dict() must be the issue.
On printing the state_dict of the runai optimizer I see
`        print(optimizer.state_dict())
`
```
{'state': {'runai': <runai.ga.torch.optim.Optimizer.<locals>._GradientAccumulationOptimizer.State object at 0x7fb6646feda0>, 0: {'momentum_buffer': tensor([[[-5.3086e-04, -1.7947e-03,  2.5590e-04,  ..., -2.6230e-04, 
...
]}
```

The 'runai' key does not containe a python dict but an object reference. This is probably why torch.save and in turn pickler.dump fails.
Should I remove the 'runai' key from the optimizer state_dict before saving?

Any clarity is appreciated.
Thank you

**Comment by levosos** [comment]
Hi @sandeepnmenon.
Thank you very much for opening this issue and for the great explanation and documentation.
I was able to reproduce the issue pretty easily thanks to your description of the issue and I appreciate it very much!

Indeed, an internal class of ours (`State`) was unpicklable and now it is.
I tested it manually and also added a test to verify it (which now pass).
I [merged](https://github.com/run-ai/runai/pull/29) the fix to `master` and published version 0.4.1 to PyPI.
Make sure you install the new version (maybe you'll have to explicitly specify `pip install runai==0.4.1`).

Please let me know if it solved the issue, and thanks again for reaching out and for this detailed description.

**Comment by sandeepnmenon** [comment]
Thank you @levosos  for the quick resolution. I am able to save the model now without having to remove the 'runai' key.
